### PR TITLE
Require C++11, always use std::function

### PR DESCRIPTION
**Pull request by zeha**
This came up in https://github.com/PowerDNS/pdns/pull/11809. PowerDNS always compiles as c++11 or newer, and the boost function pointer code is somewhat problematic on newer compiler versions (i.e. gives heaps of warnings about uninitialised memory).

As I had some trouble compiling the yahttp test suite on macOS, this also includes:
- tests: include <clocale>
- move template function body into header file instead of cpp. Otherwise, clang/ld will not see the function body.

### CODE REVIEW

**Comment by darnuria** [comment]
Did a quick check of the tests on my linux with gcc 12.1.0 and clang14.0.6

Compiled it with no option `autogen.sh && ./configure && make -j4`
then `cd tests`
```
make test                                                                                                                                                          
  CC       test-md5.o
  CXX      test-test-main.o
  CXX      test-test-md5.o
  CXX      test-test-utility.o
  CXX      test-test-url.o
  CXX      test-test-cookie.o
  CXX      test-test-request.o
  CXX      test-test-response.o
  CXX      test-test-router.o
  CXXLD    test
```

```
./test                                                                                                                                                      
Running 46 test cases...

*** No errors detected
```

Had no previous knowledge of the lib and found how to compile test with a lucky grep in Makefile so it's an indication may not have run all the test!

File: yahttp/reqresp.hpp [hunk 1]
```diff
@@ -168,9 +148,7 @@ namespace YaHTTP {
     ssize_t max_request_size; //<! maximum size of request
     ssize_t max_response_size;  //<! maximum size of response
     bool is_multipart; //<! if the request is multipart, prevents Content-Length header
-#ifdef HAVE_CPP_FUNC_PTR
-    funcptr::function<size_t(const HTTPBase*,std::ostream&,bool)> renderer; //<! rendering function
-#endif
+    std::function<size_t(const HTTPBase*,std::ostream&,bool)> renderer; //<! rendering function
```
Comment (cmouse) [inline yahttp/reqresp.hpp:151]: this commit should be split into two:

 - 1st - change the optional => mandatory
 - 2nd - do the std::function changes

**Review by cmouse — CHANGES_REQUESTED** [review]

**Comment by omoerbeek** [comment]
I tested this with PowerDNS Recusor and it does seem to play nice.

**Comment by cmouse** [comment]
> I tested this with PowerDNS Recursor and it does seem to play nice.

I am 100% happy to merge this after the mentioned commit has been split into two.

**Comment by darnuria** [comment]
@zeha Just take the liberty to ping, since all began with https://github.com/PowerDNS/pdns/pull/11809.

**Pull request closed without merging**